internal/appsec/emitter/httpsec/http.go

```diff
@@ -108,61 +108,74 @@ func makeCookies(parsed []*http.Cookie) map[string][]string {
-// HandlerOperationRes.
-// The onBlock params are used to cleanup the context when needed.
-// It is a specific patch meant for Gin, for which we must abort the
-// context since it uses a queue of handlers and it's the only way to make
-// sure other queued handlers don't get executed.
-// TODO: this patch must be removed/improved when we rework our actions/operations system
-func WrapHandler(handler http.Handler, span ddtrace.Span, pathParams map[string]string, opts *Config) http.Handler {
+func BeforeHandle(w http.ResponseWriter, r *http.Request, span ddtrace.Span, pathParams map[string]string, opts *Config) (
+	http.ResponseWriter, *http.Request, func(), bool) {
 	if opts == nil {
 		opts = defaultWrapHandlerConfig
 	} else if opts.ResponseHeaderCopier == nil {
 		opts.ResponseHeaderCopier = defaultWrapHandlerConfig.ResponseHeaderCopier
 	}
 
-	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		op, blockAtomic, ctx := StartOperation(r.Context(), HandlerOperationArgs{
-			Method:      r.Method,
-			RequestURI:  r.RequestURI,
-			Host:        r.Host,
-			RemoteAddr:  r.RemoteAddr,
-			Headers:     r.Header,
-			Cookies:     makeCookies(r.Cookies()),
-			QueryParams: r.URL.Query(),
-			PathParams:  pathParams,
-		})
-		r = r.WithContext(ctx)
-
-		defer func() {
-			var statusCode int
-			if res, ok := w.(interface{ Status() int }); ok {
-				statusCode = res.Status()
+	op, blockAtomic, ctx := StartOperation(r.Context(), HandlerOperationArgs{
+		Method:      r.Method,
+		RequestURI:  r.RequestURI,
+		Host:        r.Host,
+		RemoteAddr:  r.RemoteAddr,
+		Headers:     r.Header,
+		Cookies:     makeCookies(r.Cookies()),
+		QueryParams: r.URL.Query(),
+		PathParams:  pathParams,
+	})
+	tr := r.WithContext(ctx)
+
+	afterHandle := func() {
+		var statusCode int
+		if res, ok := w.(interface{ Status() int }); ok {
+			statusCode = res.Status()
+		}
+		op.Finish(HandlerOperationRes{
+			Headers:    opts.ResponseHeaderCopier(w),
+			StatusCode: statusCode,
+		}, span)
+
+		// Execute the onBlock functions to make sure blocking works properly
+		// in case we are instrumenting the Gin framework
+		if blockPtr := blockAtomic.Load(); blockPtr != nil {
+			for _, f := range opts.OnBlock {
+				f()
 			}
-			op.Finish(HandlerOperationRes{
-				Headers:    opts.ResponseHeaderCopier(w),
-				StatusCode: statusCode,
-			}, span)
-
-			// Execute the onBlock functions to make sure blocking works properly
-			// in case we are instrumenting the Gin framework
-			if blockPtr := blockAtomic.Load(); blockPtr != nil {
-				for _, f := range opts.OnBlock {
-					f()
-				}
-
-				if blockPtr.Handler != nil {
-					blockPtr.Handler.ServeHTTP(w, r)
-				}
+
+			if blockPtr.Handler != nil {
+				blockPtr.Handler.ServeHTTP(w, tr)
 			}
-		}()
+		}
+	}
+
+	handled := false
+	if blockPtr := blockAtomic.Load(); blockPtr != nil && blockPtr.Handler != nil {
+		// handler is replaced
+		blockPtr.Handler.ServeHTTP(w, tr)
+		blockPtr.Handler = nil
```

Comment:
This is making me feel a little unsafe (@eliottness can you have a look?).

Basically, the `blockPtr` comes from an atomic pointer, suggesting it may be concurrently modified... And then we're changing one value in there that was checked above... so we have a possible race condition here...

We may need to also use an atomic pointer here, and atomic-swap it with a `nil`?

Reply:
Took note of the issues you mention in appsec code. Although the atomic pointer was us being overzealous in the first place, let's be consistent and replace the atomic.Load() calls to atomic.CompareAndSwap in another PR so we get rid of this race